akinet/tls/client_parser_factory.go

```diff
@@ -0,0 +1,64 @@
+package tls
+
+import (
+	"github.com/akitasoftware/akita-libs/akinet"
+	"github.com/akitasoftware/akita-libs/memview"
+	"github.com/google/gopacket/reassembly"
+)
+
+// Returns a parser factory for the client half of a TLS connection.
+func NewTLSClientParserFactory() akinet.TCPParserFactory {
+	return &tlsClientParserFactory{}
+}
+
+type tlsClientParserFactory struct{}
+
+func (*tlsClientParserFactory) Name() string {
+	return "TLS 1.2/1.3 Client Parser Factory"
+}
+
+func (factory *tlsClientParserFactory) Accepts(input memview.MemView, isEnd bool) (decision akinet.AcceptDecision, discardFront int64) {
+	decision, discardFront = factory.accepts(input)
+
+	if decision == akinet.NeedMoreData && isEnd {
+		decision = akinet.Reject
+		discardFront = input.Len()
+	}
+
+	return decision, discardFront
+}
+
+func (*tlsClientParserFactory) accepts(input memview.MemView) (decision akinet.AcceptDecision, discardFront int64) {
+	if input.Len() < minTLSClientHelloLength_bytes {
+		return akinet.NeedMoreData, 0
+	}
+
+	// Accept if we match a "Client Hello" handshake message. Reject if we fail to
+	// match.
+
+	expectedBytes := map[int]byte{
+		// Record header (5 bytes)
+		0: 0x16,          // handshake record
+		1: 0x03, 2: 0x01, // protocol version 3.1 (TLS 1.0)
+		// 2 bytes of handshake payload size
+
+		// Handshake header (4 bytes)
+		5: 0x01, // Client Hello
+		// 3 bytes of Client Hello payload size
+
+		// Client Version (2 bytes)
+		9: 0x03, 10: 0x03, // protocol version 3.3 (TLS 1.2)
+	}
+
+	for idx, expectedByte := range expectedBytes {
+		if input.GetByte(int64(idx)) != expectedByte {
+			return akinet.Reject, input.Len()
+		}
+	}
```

Comment:
OK, I see what you did there.

I think the first problem is that expectedBytes should be at package level. Go will create a new map on each call to parse, when we should reuse the same one. 

I did a benchmark: https://gist.github.com/mgritter/62157436f582569add1ea24b5088ba22
234ns for creating the map inside the loop, vs. 73ns for outside

(Similar results for having it be a package-level var.)

We can get an even bigger win by masking a byte array instead of a map:

```
	expected := []byte{0x16, 0x03, 0x01, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x03, 0x03}
	mask := []byte{0xff, 0xff, 0xff, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0xff, 0xff}

	for n := 0; n < b.N; n++ {
		result = true
		for i := range expected {
			if (mv.GetByte(int64(i)) & mask[i]) != expected[i] {
```

That gets the cost of a successful check down to 12-13 ns.  Which seems worth it because this check is run every time we have unknown bytes we're trying to determine a parser for?

Reply:
In 3d7e5e9.